src/ripple/protocol/impl/STTx.cpp

```diff
@@ -307,13 +308,48 @@ isMemoOkay (STObject const& st, std::string& reason)
 
         for (auto const& memoElement : *memoObj)
         {
-            if ((memoElement.getFName() != sfMemoType) &&
-                (memoElement.getFName() != sfMemoData) &&
-                (memoElement.getFName() != sfMemoFormat))
+            auto const& name = memoElement.getFName();
+
+            if (name != sfMemoType &&
+                name != sfMemoData &&
+                name != sfMemoFormat)
+            {
+                reason = "A memo may contain only MemoType, MemoData or "
+                         "MemoFormat fields.";
+                return false;
+            }
+
+            // The raw data is stored as hex-octets, which we want to decode.
+            auto data = strUnHex (memoElement.getText ());
+
+            if (!data.second)
             {
-                reason = "A memo may contain only MemoType, MemoData or MemoFormat fields.";
+                reason = "The MemoType, MemoData and MemoFormat fields may "
+                         "only contain hex-encoded data.";
                 return false;
             }
+
+            if (name == sfMemoData)
+                continue;
+
+            // The only allowed characters for MemoType and MemoFormat are the
+            // characters allowed in URLs per RFC 3986: alphanumerics and the
+            // following symbols:
+            static std::string const allowed_symbols ("-._~:/?#[]@!$&'()*+,;=%");
+
+            for (auto c : data.first)
+            {
+                if (isalnum (c))
```

Comment:
We can remove the dependency on `isalnum` fairly easily actually. This might also marginally improve the code that tests for symbols by doing a range check. Something like:

    if (c >= 'A' && c <= 'Z' || 
        c >= 'a' && c <= 'z' || 
        c >= '0' && c <= '9')